### ISSUE TITLE
Easier and more intuitive scaling for simple shapes.

### ISSUE THREAD

**Issue by Aubron** [fabricjs/api-discussion#7]
## The problem to solve

Presently, if you use drag handles to 'scale' a Rect, polygon, etc, its internal height and width stay the same. This is fine, except in situations where you want to say, add a stroke. It will be added incongruently and inconsistently if the shape is non-uniformly scaled. As well, setting a stroke of '5' on a object with a scale of 3x is actually setting a stroke of 15.

## Current situation
We are implementing post modification edits to resize objects based on the scale provided. This works fine, but is a lot of boilerplate code, especially for complex objects like Poly{things}.

## Proposal
This kind of functionality feels like it should be the default, or at least a toggleable option.

## Breaking changes
A lot. I would estimate a lot of projects include this boilerplate code, and it would potentially malfunction and double or triple scale the object. Could be avoided by making it a toggle on instead of a toggle off.

**Comment by asturur** [comment]
So this feature has been requested for ages.
A coworker of mine sort of made a prototype that is not exactly what you are talking about but is similar.

First of all i want to highlight that SVG behaves as fabricJS ( or vice versa ) a scaling is a scaling, and should scale the stroke too.

There are 2 options to keep in mind here.

First option:

You can either change the size of the objects as you are doing, in this case the boiler plate code would be sort of embedded in fabricJS but will not be less
The code is potentially very slow for complex paths, and need to be executed on each frame of scaling, and invalidates the cache each frame of scaling, it can't be optional.

Second option:

Do not resize objects, add an option to represent stroke always has object had scale of 1. This is much easier to implement and requires less coding. It requires anyway to rewrite the logic of bounding box for objects, it has the same drawback on cache, needs an invalidation per frame of scaling to be visually consistent. ( it can be left as an option )

Other things to consider:
Having the ability to make a path abosrb a transformation and get the new points of the path ( or poly{thing}) is anyway a wanted feature and should be implemented.

Customizable controls is something else that people are looking forward to, and instead of mess up the scaling logic with a resizing one, it would be nice if devs could add their own handlers to control in a way the original logic is disconnected and a new one can be used.
Rewriting the actual rotation/scaling logic as an handler would be required and would serve as an example on how to build your own, for example, resizing logic.

Groups:
When scaling a group i can assure you that scaling the stroke with it is really the only option, or otherwise anything that looks like a composition gets screwed up. Also in this case resizing is sort of harder. In this case i would not even know what is the best thing to do.

In some way we will have consistent stroke during scaling, probably after clipPath have landed and canvas rotation is setup for a try, this is definitely a discussion i would like to bring on.
@stefanhayden is someone could have an opinion about it, and also @fabricjs/api-discussion involved in the group.

**Comment by Aubron** [comment]
Again, we're probably taking an over simplistic approach here, but for us and groups, the same logic applies ('it should line up with what the cached scale looks like before mouseup, with the exception of stroke and whatnot')

**Comment by Aubron** [comment]
On further inspection I am seeing some issues with how we do this. If a complex object already has a stroke, it jumps on modification, as taking into account the bounding box of the changing stroke is prohibitively difficult. For rectangle it looks like we take into account the stroke when calculating the width and height, but I don't really want to think about anything more complex than that.

**Comment by asturur** [comment]
As a side not, offtopic from the discussion, the positioning problems are generally solved with setPositionByOrigin: on your mouse up event, take the object.getCenterPoint() as first thing, do what you want to the object, then do object.setPositionByOrigin(oldCenter, 'center', 'center') and will probably stop jumping.
But let's keep the offtopic as minimal as possible.

**Comment by stefanhayden** [comment]
yeah I would go for changing the stroke as the object scales. The math for this is fairly simple. You can do it live as the object scales though you obviously loose some performance.

This breaks svg compatibility as well but in many situations that's not really a big deal.

**Comment by blobinabottle** [comment]
Note regarding svg compatibility you can use vector-effect="non-scaling-stroke"
